WIP: Use ffmpeg when sox binary is missing

hagenw opened the pull request:
Closes #62.

This falls back to `ffmpeg` if the `sox` binary is missing on the system as `pysox` does not ensure the binary to be present.

In addition, it adds tests to make sure the error messages a user gets for broken or missing files are the same if the `sox` binary is missing or not.

I also updated the installation instructions, highlighting that `sox` is only needed for faster MP3 file handling:

![image](https://user-images.githubusercontent.com/173624/162899450-90c73c80-f8c3-4ac6-953a-99ec7fe2e4b2.png)

frankenjoe commented:
Nice tests! The `indirect` argument to pass values to fixtures was new to me, very convenient.

frankenjoe commented on tests/test_symlink.py:
```diff
@@ -0,0 +1,19 @@
+    print(os.environ['PATH'])
+    os.environ['PATH'] = path
+    print(os.environ['PATH'])
+    assert shutil.which(program) == local_command
```
Maybe the copying solution does not work due to missing dependencies? We probably have to copy some shared libraries as well.

frankenjoe replied:
Maybe the copying solution does not work due to missing dependencies? We probably have to copy some shared libraries as well.

hagenw replied:
Locally it works for me when doing it with Unix commands.
In the Github runner it does not fail when executing `ffmpeg`, but it cannot find it in the path: https://github.com/audeering/audiofile/runs/5756308827?check_suite_focus=true

frankenjoe replied:
Did you try with `.exe`?

frankenjoe replied:
According to https://stackoverflow.com/questions/56399696/shutil-which-not-finding-programs-without-appending-file-extension: "shutil.which('ffmpeg') will look for "ffmpeg.EXE" in PATH", but I guess the executable is called `ffmpeg.exe`.

hagenw replied:
Renaming it to `ffmpeg.EXE` did indeed help and the custom `tests/test_symlink.py` no works.

With the actual audio tests we still get 4 failures under Windows for which I'm not sure why yet: https://github.com/audeering/audiofile/runs/5770153335?check_suite_focus=true

frankenjoe replied:
I guess best would be to run it on a Windows machine locally and debug what's going on. I can later give it a try on my wife's machine if you want.

hagenw replied:
That would be nice.
If you don't need this fix immediately it can also wait for a couple of days.

frankenjoe replied:
At least locally tests are passing on Windows:

![image](https://user-images.githubusercontent.com/10383417/161398134-43f18ecf-2007-4e17-830e-610668b850a7.png)

So maybe we simply skip the failing tests on Windows?

hagenw replied:
I'm not a big fan of skipping the tests. But as we seem to need a fix as soon as possible and we don't have time to first investigate https://github.com/audeering/audiofile/issues/65 it might be ok for an intermediate release to skip the test.

I will take a look and update the code accordingly.

hagenw replied:
On the other hand this problem would disappear when we just remove `sox`.

I started running the benchmarks with and without `sox`, let's first see what the outcome will be.

hagenw replied:
I solved it in https://github.com/audeering/audiofile/pull/69 by using Github runners with and without `sox` installed. Then we don't need to mess around with the path as we do here.

frankenjoe replied:
Hehe, I also thought about using different Github runners, but didn't want to suggest it after you already spend to effort on hiding `sox` from PATH :)

frankenjoe commented:
Looking at the benchmark for reading mp3/mp4, `audioread` using `ffmpeg` (?) seems not to be slower than `audiofile` using `sox` (?). If that is true, why do we need `sox` at all?

![image](https://user-images.githubusercontent.com/10383417/162149981-9c6ff398-6fa3-4a51-8b74-0c91473fd24d.png)

hagenw commented:
`sox` is more lightweight than `ffmpeg` and on some systems it supports MP3. I think this was the main motivation to include it.

hagenw commented:
For `audioread` it is not easy to predict what it is using:

![image](https://user-images.githubusercontent.com/173624/162688190-dcb3fa4d-a278-4837-b1cf-ee93be9844a7.png)

In addition, it can be slow at accessing metadata:

![image](https://user-images.githubusercontent.com/173624/162688281-6410b0ad-7ee8-404a-8a90-99277d4f7687.png)

So, in general I was not very happy using it.

hagenw commented:
> `sox` is more lightweight than `ffmpeg` and on some systems it supports MP3. I think this was the main motivation to include it.

It might be that it is faster in reading MP3 files and accessing it's metadata than `ffmpeg` and `mediainfo`. We should maybe test that. If it is not the case we might indeed remove `sox` completely.

frankenjoe commented:
> So, in general I was not very happy using it.

I'm not suggesting to use `audioread`, but get rid of `sox` and always use `ffmpeg`. This also has the advantage that the result does not depend on whether `sox` is installed or not.

hagenw commented:
I created https://github.com/audeering/audiofile/issues/65 to track the issue of removing `sox` altogether.

hagenw commented:
Looking at the results from https://github.com/audeering/audiofile/issues/65#issuecomment-1095209160 I would be in favor of sticking with `sox`. To handle the failing test, we first need to merge https://github.com/audeering/audiofile/pull/68 where I switch to `conda` for installing `sox`.

I also updated the installation instructions accordingly and added it to the description of this pull request.

hagenw commented on audiofile/core/convert.py:
```diff
@@ -16,6 +14,11 @@
 
+# Import sox after disabling warning
+# as it is ok to not have the binary present
+import sox  # pragma: E402
```
Moving the import os `sox` after we disable the sox warnings should solve the issue of showing the warning (and failing building docs when `sox` is not installed). The problem is that I get a flake8 E402 error here even though I use `pragma`, or do I something wrong?

hagenw commented:
Closing in favor of #69.
